### ISSUE TITLE
Sub document metadata issue

### ISSUE THREAD

**Issue by heylookalive** [typegoose/typegoose#283]
## Versions
- NodeJS: 12.3.1
- Typegoose(NPM): 7.1.3
- mongoose: 5.9.16
- grahpql / type-graphql: 15.0.0

## What is the Problem?

Using Typegoose with TypeGraphQL I've got some types which have sub documents and references for what will be a complex graph.

My problem is that when referencing a sub document which is called "meta" Typegoose is failing to find the metadata for this property. I'm not sure what I'm getting wrong here. If I move all the type classes into the same file then the issue doesn't manifest. Any help would be appreciated.

## Code Example

I've setup a basic repo where the issue can be replicated - https://github.com/heylookalive/typegoose-issue-example

```ts
[nodemon] starting `ts-node src/index.ts`

/Users/xx/Sites/typegoose-test/node_modules/reflect-metadata/Reflect.js:553
                var decorated = decorator(target, propertyKey, descriptor);
                                ^
Error: There is no metadata for the "meta" property.
Check if emitDecoratorMetadata is enabled in tsconfig.json or check if you've declared a sub document's class after usage.
    at /Users/xx/Sites/typegoose-test/node_modules/@typegoose/typegoose/src/prop.ts:383:53
    at DecorateProperty (/Users/xx/Sites/typegoose-test/node_modules/reflect-metadata/Reflect.js:553:33)
    at Object.decorate (/Users/xx/Sites/typegoose-test/node_modules/reflect-metadata/Reflect.js:123:24)
    at __decorate (/Users/xx/Sites/typegoose-test/src/entities/User.ts:4:92)
    at /Users/xx/Sites/typegoose-test/src/entities/User.ts:13:2
    at Object.<anonymous> (/Users/xx/Sites/typegoose-test/src/entities/User.ts:33:3)
    at Module._compile (internal/modules/cjs/loader.js:774:30)
    at Module.m._compile (/Users/xx/Sites/typegoose-test/node_modules/ts-node/src/index.ts:858:23)
    at Module._extensions..js (internal/modules/cjs/loader.js:785:10)
    at Object.require.extensions.<computed> [as .ts] (/Users/xx/Sites/typegoose-test/node_modules/ts-node/src/index.ts:861:12)
    at Module.load (internal/modules/cjs/loader.js:641:32)
    at Function.Module._load (internal/modules/cjs/loader.js:556:12)
    at Module.require (internal/modules/cjs/loader.js:681:19)
    at require (internal/modules/cjs/helpers.js:16:16)
    at Object.<anonymous> (/Users/xx/Sites/typegoose-test/src/entities/Common.ts:3:1)
    at Module._compile (internal/modules/cjs/loader.js:774:30)
[nodemon] app crashed - waiting for file changes before starting...
```

## Do you know *why* it happens?

No.

**Comment by andreialecu** [comment]
I suspect this is because of the circular dependency between `User.ts` and `Common.ts`.

If you move the `Resolvers` to a separate file everything will work.

It would be interesting to see if deferring this to be called async could fix the issue:
https://github.com/typegoose/typegoose/blob/a660ab66fe3e544dc00c9c300b5e7f511881f1ec/src/prop.ts#L456-L462

Basically just a `setImmediate()` wrapper around it. Postponing the actions executed in the decorator could work. cc @hasezoey

**Comment by hasezoey** [comment]
@andreialecu adding any async thing to an decorator would mean there is an possibility for race-conditions, because an decorator cannot be awaited, and when it would need to wrap the whole decorator and not just the call to `baseProp` (error is thrown on line 383, before `baseProp` call) https://github.com/typegoose/typegoose/blob/a660ab66fe3e544dc00c9c300b5e7f511881f1ec/src/prop.ts#L383

**Comment by andreialecu** [comment]
Another thing that works to fix your repo is to reorder these two imports in `index.ts`:

```ts
import { UserResolver } from "./entities/User";
import { OrganisationResolver } from "./entities/Org";
```

`index.ts` imports both `User` and `Org`. Then each of `User` and `Org` imports `Common`, and `Common` imports `User`.

The order of imports seems to be important, but as a better practice you should remove all circular references from your projects. 

As a good practice, it's good to enforce this rule: 
https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/no-cycle.md
https://spin.atomicobject.com/2018/06/25/circular-dependencies-javascript/

**Comment by andreialecu** [comment]
@hasezoey since the error comes from the assertion, from what I can see is that it tries to read the `design:type` of the property. But since the type can be supplied by `@prop(type => SomeType)` is the assertion really necessary?

**Comment by heylookalive** [comment]
I do agree on circular dependencies but it's required to produce the graph as wanted.

All in TS and typegraphql are handling the fact there's circular deps but typegoose is having trouble, and the one throwing the error.

I can see that there's provision for this sort of thing for references where you can mention a class name via string, perhaps this is needed for embedded docs too.
```
@prop({ ref: 'User' })
```

**Comment by hasezoey** [comment]
@andreialecu `design:type` is used for primitives and nowdays to detect maps/arrays, but its true, undefined should be an allowed one, as long as rawOptions.type is supplied (i guess i will change it so that it dosnt error on undefined `Type` value, but when `Type` **and** `rawOptions.type` are undefined/null)

**Comment by hasezoey** [comment]
this should be fixed in ~7.3

@heylookalive could you test again with `v7.3.0-1`?

**Comment by hasezoey** [comment]
closing because this should be fixed since 7.3
